// @deskthing/types/dist/clients/clientTransit.js
/**
 * You generally do not have to worry about this
 */
export var CLIENT_REQUESTS;
(function (CLIENT_REQUESTS) {
    CLIENT_REQUESTS["GET"] = "get";
    CLIENT_REQUESTS["ACTION"] = "action";
    CLIENT_REQUESTS["BUTTON"] = "button";
    CLIENT_REQUESTS["KEY"] = "key";
    CLIENT_REQUESTS["LOG"] = "log";
})(CLIENT_REQUESTS || (CLIENT_REQUESTS = {}));
/**
 * You generally do not have to worry about this
 *
 */
export var DEVICE_CLIENT;
(function (DEVICE_CLIENT) {
    DEVICE_CLIENT["MANIFEST"] = "manifest";
    DEVICE_CLIENT["MUSIC"] = "music";
    DEVICE_CLIENT["SETTINGS"] = "settings";
    DEVICE_CLIENT["APPS"] = "apps";
    DEVICE_CLIENT["ACTION"] = "action";
    DEVICE_CLIENT["TIME"] = "time";
    DEVICE_CLIENT["ICON"] = "icon";
})(DEVICE_CLIENT || (DEVICE_CLIENT = {}));
export var DEVICE_DESKTHING;
(function (DEVICE_DESKTHING) {
    DEVICE_DESKTHING["MANIFEST"] = "manifest";
    DEVICE_DESKTHING["ACTION"] = "action";
    DEVICE_DESKTHING["SET"] = "set";
    DEVICE_DESKTHING["PING"] = "ping";
    DEVICE_DESKTHING["PONG"] = "pong";
    DEVICE_DESKTHING["SETTINGS"] = "settings";
    DEVICE_DESKTHING["LOG"] = "log";
    DEVICE_DESKTHING["VIEW"] = "view";
    DEVICE_DESKTHING["APP_PAYLOAD"] = "app_payload";
})(DEVICE_DESKTHING || (DEVICE_DESKTHING = {}));
