// @deskthing/types/dist/client.js
export var ClientConnectionMethod;
(function (ClientConnectionMethod) {
    ClientConnectionMethod[ClientConnectionMethod["Unknown"] = 0] = "Unknown";
    ClientConnectionMethod[ClientConnectionMethod["LAN"] = 1] = "LAN";
    ClientConnectionMethod[ClientConnectionMethod["Localhost"] = 2] = "Localhost";
    ClientConnectionMethod[ClientConnectionMethod["ADB"] = 3] = "ADB";
    ClientConnectionMethod[ClientConnectionMethod["NDIS"] = 4] = "NDIS";
    ClientConnectionMethod[ClientConnectionMethod["Bluetooth"] = 5] = "Bluetooth";
    ClientConnectionMethod[ClientConnectionMethod["Internet"] = 6] = "Internet";
})(ClientConnectionMethod || (ClientConnectionMethod = {}));
export var ClientPlatformIDs;
(function (ClientPlatformIDs) {
    ClientPlatformIDs[ClientPlatformIDs["Unknown"] = 0] = "Unknown";
    ClientPlatformIDs[ClientPlatformIDs["Desktop"] = 1] = "Desktop";
    ClientPlatformIDs[ClientPlatformIDs["Tablet"] = 2] = "Tablet";
    ClientPlatformIDs[ClientPlatformIDs["Iphone"] = 3] = "Iphone";
    ClientPlatformIDs[ClientPlatformIDs["CarThing"] = 4] = "CarThing";
})(ClientPlatformIDs || (ClientPlatformIDs = {}));
/**
 * @module deskthing/client
 * @description
 * Volume mode enum used for configuring volume control behavior
 * @see {@link ClientPreferences.volume}
 */
export var VolMode;
(function (VolMode) {
    VolMode["WHEEL"] = "wheel";
    VolMode["SLIDER"] = "slider";
    VolMode["BAR"] = "bar";
})(VolMode || (VolMode = {}));
/**
 * @module deskthing/client
 * @description
 * View mode enum used for configuring display states
 * @see {@link ClientPreferences.appTrayState}
 * @see {@link MiniplayerSettings.state}
 */
export var ViewMode;
(function (ViewMode) {
    ViewMode["HIDDEN"] = "hidden";
    ViewMode["PEEK"] = "peek";
    ViewMode["FULL"] = "full";
})(ViewMode || (ViewMode = {}));
/**
 * You generally do not have to worry about this
 */
export var ToDeviceDataEvents;
(function (ToDeviceDataEvents) {
    ToDeviceDataEvents["GET"] = "get";
    ToDeviceDataEvents["ACTION"] = "action";
    ToDeviceDataEvents["BUTTON"] = "button";
    ToDeviceDataEvents["KEY"] = "key";
    ToDeviceDataEvents["LOG"] = "log";
})(ToDeviceDataEvents || (ToDeviceDataEvents = {}));
/**
 * You generally do not have to worry about this
 */
export var FromDeviceDataEvents;
(function (FromDeviceDataEvents) {
    FromDeviceDataEvents["MANIFEST"] = "manifest";
    FromDeviceDataEvents["MUSIC"] = "music";
    FromDeviceDataEvents["SETTINGS"] = "settings";
    FromDeviceDataEvents["APPS"] = "apps";
    FromDeviceDataEvents["ACTION"] = "action";
    FromDeviceDataEvents["TIME"] = "time";
    FromDeviceDataEvents["ICON"] = "icon";
})(FromDeviceDataEvents || (FromDeviceDataEvents = {}));
/**
 * All data that is sent from the DeskThing Server to the Client
 * @package DeskThingServer + DeskThingClient
 */
export var FromDeskthingToDeviceEvents;
(function (FromDeskthingToDeviceEvents) {
    /** Will never be emitted to your client. Only exists on device */
    FromDeskthingToDeviceEvents["GLOBAL_SETTINGS"] = "global_settings";
    /** Will never be emitted to your client. Only exists on device */
    FromDeskthingToDeviceEvents["MAPPINGS"] = "button_mappings";
    /** Will never be emitted to your client. Only exists on device */
    FromDeskthingToDeviceEvents["GET"] = "get";
    /** Will never be emitted to your client. Only exists on device */
    FromDeskthingToDeviceEvents["ERROR"] = "error";
    FromDeskthingToDeviceEvents["PONG"] = "pong";
    FromDeskthingToDeviceEvents["PING"] = "ping";
    FromDeskthingToDeviceEvents["HEARTBEAT"] = "heartbeat";
})(FromDeskthingToDeviceEvents || (FromDeskthingToDeviceEvents = {}));
