// @deskthing/types/dist/music.js
/**
 * Various audio requests that will be sent to an audio source
 */
export var AUDIO_REQUESTS;
(function (AUDIO_REQUESTS) {
    AUDIO_REQUESTS["NEXT"] = "next";
    AUDIO_REQUESTS["PREVIOUS"] = "previous";
    AUDIO_REQUESTS["REWIND"] = "rewind";
    AUDIO_REQUESTS["FAST_FORWARD"] = "fast_forward";
    AUDIO_REQUESTS["PLAY"] = "play";
    AUDIO_REQUESTS["PAUSE"] = "pause";
    /** @depreciated */
    AUDIO_REQUESTS["STOP"] = "stop";
    AUDIO_REQUESTS["SEEK"] = "seek";
    AUDIO_REQUESTS["LIKE"] = "like";
    AUDIO_REQUESTS["SONG"] = "song";
    AUDIO_REQUESTS["VOLUME"] = "volume";
    AUDIO_REQUESTS["REPEAT"] = "repeat";
    AUDIO_REQUESTS["SHUFFLE"] = "shuffle";
    /** Only used by the server */
    AUDIO_REQUESTS["REFRESH"] = "refresh";
})(AUDIO_REQUESTS || (AUDIO_REQUESTS = {}));
/**
 * A specific SongEvent enum for filtering only song-related events
 */
export var SongEvent;
(function (SongEvent) {
    /**
     * request: 'song' | 'refresh'
     */
    SongEvent["GET"] = "get";
    /**
     * request: "next" | "previous" | "fast_forward" | "rewind" | "play" | "pause" | "stop" | "seek" | "like" | "volume" | "repeat"
     */
    SongEvent["SET"] = "set";
})(SongEvent || (SongEvent = {}));
