// @deskthing/types/dist/utils.js
export var LOGGING_LEVELS;
(function (LOGGING_LEVELS) {
    LOGGING_LEVELS["MESSAGE"] = "message";
    LOGGING_LEVELS["LOG"] = "log";
    LOGGING_LEVELS["WARN"] = "warning";
    LOGGING_LEVELS["ERROR"] = "error";
    LOGGING_LEVELS["DEBUG"] = "debugging";
    LOGGING_LEVELS["FATAL"] = "fatal";
})(LOGGING_LEVELS || (LOGGING_LEVELS = {}));
/**
 * Different supported platforms for the app
 */
export var PlatformTypes;
(function (PlatformTypes) {
    PlatformTypes["WINDOWS"] = "windows";
    PlatformTypes["LINUX"] = "linux";
    PlatformTypes["MAC"] = "mac";
    PlatformTypes["MAC64"] = "mac64";
    PlatformTypes["MACARM"] = "macarm";
    PlatformTypes["ANDROID"] = "android";
    PlatformTypes["IOS"] = "ios";
    PlatformTypes["ARM64"] = "arm64";
    PlatformTypes["X64"] = "x64";
})(PlatformTypes || (PlatformTypes = {}));
/**
 * For the manifest. Different types of tags that can be used to categorize apps
 */
export var TagTypes;
(function (TagTypes) {
    TagTypes["AUDIO_SOURCE"] = "audiosource";
    TagTypes["SCREEN_SAVER"] = "screensaver";
    TagTypes["UTILITY_ONLY"] = "utilityOnly";
    TagTypes["WEB_APP_ONLY"] = "webappOnly";
})(TagTypes || (TagTypes = {}));
